### DISCUSSION TITLE
Node IDs

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#21] [General]
Currently, node IDs are either integers or strings.
* https://pg-format.readthedocs.io/en/0.3/contents/pg-format.html

An idea came up to restrict node IDs to strings only.
However, if this idea is applied, JSON examples previously presented in the paper will become invalid.
* https://arxiv.org/abs/1907.03936

**Comment by hchiba1** [comment]
Actually, the previous PG parser has an issue in this regard: it does not retain the data type of parsed literals, causing integers to be converted into strings.
In the refined implementation, the parser retains data types of parsed literals, so integers are no longer converted into strings.

**Reply by nichtich** [reply]
Node identifiers, labels and property names can always be converted to strings, while property values have data type, so the PG format graph

    1 :2 3:4 3:"5"

is parsed as:

~~~json
{"id":"1","labels":["2"],"properties":{"3":[4,"5"]}}
 ~~~

**Comment by nichtich** [comment]
Quoted node ids can be anything, even the empty string. Unquoted node ids, labels, and property keys should be at least allow for these use cases:

- plain numbers
- alphanumeric strings
- strings like `label@de`, `dc:title`...
- all kinds of URIs
- all kinds of IRI

I tried to find out by reading RFC and testing IRI libraries what characters are disallowed in IRIs, this is a bit tricky, but it looks like this from the spec is relevant:

> Systems accepting IRIs MAY also deal with the printable characters in
   US-ASCII that are not allowed in URIs, namely "<", ">", '"', space,
   "{", "}", "|", "\\", "^", and "`", in step 2 above.

We could write down some guideline what unquoted identifier SHOULD be but be very lax in what it MUST be (disallow space and quotation mark any maybe some of the characters above but not more).

**Comment by nichtich** [comment]
As `|` is disallowed in unquoted values anyway, I've changed the additional line separator from `;` to `|`. See https://github.com/pg-format/pgraphs/blob/main/docs/pg-grammar.md for updated grammar.

**Comment by hchiba1** [comment]
I have an impression that some users may try to use integers as node IDs.

**Comment by hchiba1** [comment]
What is the drawback of allowing integers as node IDs?

Maybe users who want to restrict to only strings cannot write as:
```
1 -> 2
```
but
```
“1" -> “2"
```
will work.

This may seem a little cumbersome, but it is similar to the behavior of regular programming languages.

**Comment by nichtich** [comment]
> What is the drawback of allowing integers as node IDs?

This would be confusing because then we can have two distinct nodes with same id when printed without quotation marks or other kind of markup. (`1 -> "1"` would be an edge connecting two different nodes). I don't know of any format or graph database system that allows to have two nodes, one with id `1` as number and the other with id `1` as string. We need to differentiate:

- On the abstract data model node IDs should only be non-empty strings (Cypher further disallows NULL-Bytes).
- In PG format serializations these node IDs can be written in multiple forms, e.g. `1`, `"1"`, `"\u0031"`.
- Backend systems that PG is converted from/to such as databases and other formats have their own restrictions on node identifiers and may convert them to numbers if needed.

**Comment by hchiba1** [comment]
Basically, I agree with limiting node IDs to strings, given that the strings can practically accommodate various types of data and can be mapped to numbers if needed.

Also, I suppose there are still ways to salvage legacy PG-JSON data containing integer node IDs in practice. 
In fact, such data may be easily parsed by naive JavaScript programs that do not strictly validate the schema.

**Reply by nichtich** [reply]
I've added a non-normative section [robustness-principle](https://github.com/pg-format/specification?tab=readme-ov-file#robustness-principle) to the specification draft, also mentioning numeric identifiers.
